blogposts/2022/introducting-migrator-service.md

```diff
@@ -0,0 +1,148 @@
+3. We deploy to Cloud by starting new services that immediately attempt to run migrations.
+
+The first observation serves as loose evidence of a correlation between concurrent index creation and dirty database issues. This correlation does, albeit unscientifically, match reality.
+
+The second observation implies that we never actually check the _error message_ that occurs when the database was first marked as dirty. Looking back, a number of downtime incidents we originally attributed to migrations not completing within our Kubernetes health check timeout are also explainable by this deadlock condition. We've been increasingly relying on concurrent index creation, and _any_ contention of a migration advisory lock will cause the deadlock to occur. And, by the third observation, we have plenty of lock contention when multiple pods start up at the same time looking to update the database.
+
+With a new theory in hand, we can finally attack the problem in a focused manner.
+
+We changed the migration process to gracefully deals with concurrent index creation by dropping the advisory lock and instead coordinating with concurrent migration attempts within the Postgres catalog itself. Of particular interest is the `pg_stat_progress_create_index` system view, which [reports index creation progress](https://www.postgresql.org/docs/12/progress-reporting.html). The migrator is also intelligence enough to detect invalid indexes are repair them rather than flagging a human to hold its hand.
+
+After another round of testing, migrations finally behave according to intuition.
+
+## Conclusion
+
+We've long since outgrown the way in which we run database migrations. We've known for a while that there _was_ a problem, and felt the pain along with our users for an embarrassingly long time. After digging a bit deeper, better understand the problem space, and breaking a bit of our ossified code and deployment process along the way, I hope we're delivering a quality of life improvement to Sourcegraphers and our users.
+
+If your next upgrade does **not** crash, please tweet me at [@ericfritz](https://twitter.com/ericfritz).
+
+### Related resources
+
+We've updated several documentation pages which are worth a look for existing site administrators:
+
+- [How to run `migrator` operations](https://docs.sourcegraph.com/admin/how-to/manual_database_migrations)
+- [How to troubleshoot a dirty database](https://docs.sourcegraph.com/admin/how-to/dirty_database) (replacing the [legacy documentation](https://docs.sourcegraph.com/admin/how-to/dirty_database_pre_3_37))
+
+We are open source! Check out some of the higher-scope code changes that made this change possible:
```

Comment:
Is there a better term or approved copy to use for things like this?

Reply:
I'm not sure! Will check

Reply:
Didn't want to attribute open-core to a non-(truly)-open-core tech. Seems like being precise with what we identify ourselves as here is necessary to prevent some bad opinions from forming when we post to places like HN.

Reply:
Agreed, I think precision is key. Kacie said we're not really open source, so I'm wondering if it's better to say something like "We support open source!" We could also omit that statement and just tell readers they can check out the code changes.

Reply:
```suggestion
Check out some of the higher-scope code changes that made this change possible:
```